flashgbx.1
.TH FlashGBX 1
.SH NAME
FlashGBX \- Utility for reading and writing Game Boy and Game Boy Advance cartridge data.
.SH SYNOPSIS
.B flashgbx
[\fIOPTIONS\fR] [\fIfile\fR]
.SH DESCRIPTION
FlashGBX is a utility for reading and writing ROM, save and RTC data from your Game Boy and Game Boy Advance cartridges. It has both CLI and GUI functionality. It features functionality for exporting GB Camera photos. It queries the chip on the cartiridge to check for compatibility and has a wide range of compatible chips.
.SH OPTIONS
.SS General Arguments
.TP
.BR \-\-cli
Force the command line interface mode.
.TP
.BR \-\-reset
Clears all settings such as last used directory information
.TP
.BR \-\-debug
Enable debug messages used for development
.SS Configuration Arguments
.TP
.BR \-\-cfgdir " \fR[\fPappdata\fR|\fPsubdir\fR]"
Sets the config directory to either the OS-provided local app config directory or a subdirectory of this application
.SS Main Command Arguments
.TP
.BR \-\-mode " \fR[\fPdmg\fR|\fPagb\fR]"
Set cartridge mode to \"dmg\" (Game Boy) or \"agb\" (Game Boy Advance)
.TP
.BR \-\-action " \fR[\fPinfo\fR|\fPbackup-rom\fR|\fPbackup-save\fR|\fPrestore-save\fR|\fPerase-save\fR|\fPgbcamera-extract\fR|\fPfwupdate-gbxcartrw\fR|\fPdebug-test-save\fR]"
Select program action
.TP
.BR \-\-overwrite
Overwrite without asking if target file already exists
.SS Optional Command Arguments
.TP
.BR \-\-dmg\-romsize " \fR[\fPauto\fR|\fP32kb\fR|\fP64kb\fR|\fP128kb\fR|\fP256kb\fR|\fP512kb\fR|\fP1mb\fR|\fP2mb\fR|\fP4mb\fR|\fP8mb\fR|\fP16mb\fR|\fP32mb\fR]"
Set the size of the Game Boy cartridge ROM data
.TP
.BR \-\-dmg\-mbc " \fR[\fPauto\fR|\fP1\fR|\fP2\fR|\fP3\fR|\fP4\fR|\fP5\fR|\fP6\fR|\fP7\fR]"
Set the memory bank controller type of the Game Boy Cartridge
.TP
.BR \-\-dmg\-savesize " \fR[\fPauto\fR|\fP4k\fR|\fP16k\fR|\fP64k\fR|\fP256k\fR|\fP512k\fR|\fP1m\fR|\fPeeprom2k\fR|\fPeeprom4k\fR|\fPtama5\fR|\fP4m\fR]"
Set the Game Boy cartridge save data type
.TP
.BR \-\-agb\-romsize " \fR[\fPauto\fR|\fP4mb\fR|\fP8mb\fR|\fP16mb\fR|\fP32mb\fR|\fP64mb\fR|\fP128mb\fR|\fP256mb\fR]"
Set the size of the Game Boy Advance cartridge ROM data
.TP
.BR \-\-agb\-savetype " \fR[\fPauto\fR|\fPeeprom4k\fR|\fPeeprom64k\fR|\fPsram256k\fR|\fPflash512k\fR|\fPflash1m\fR|\fPdacs8m\fR|\fPsram512k\fR|\fPsram1m\fR]"
Set the Game Box Advance save data type
.TP
.BR \-\-store\-rtc
Store RTC register values if supported
.TP
.BR \-\-ignore\-bad\-header
Don't stop if invalid data found in cartridge header data
.TP
.BR \-\-flashcart\-type " \fR[\fItype\fP]"
Name of the flash cart, see txt files in config directory for
.I type
options
.TP
.BR \-\-prefer\-chip\-erase
Prefer full chip erase over sector erase when both available
.TP
.BR \-\-reversed\-sectors
Use reversed flash sectors if possible
.TP
.BR \-\-force\-5v
Force 5V when writing Game Boy flash cartridges
.TP
.BR \-\-no\-verify\-write
Do not verify written data
.TP
.BR \-\-save\-filename\-add\-datetime
Adds a timestamp to the file name of the save data backups
.TP
.BR \-\-gbcamera\-palette " \fR[\fPgrayscale\fR|\fPdmg\fR|\fPsgb\fR|\fPcgb1\fR|\fPcgb2\fR|\fPcgb3\fR]"
Sets the palette of the pictures extracted from the Game Boy Camera saves
.TP
.BR \-\-gbcamera\-outfile\-format " \fR[\fPpng\fR|\fPbmp\fR|\fPgif\fR|\fPjpg\fR]"
Sets the file format of the saved pictures from the Game Boy Camera save
.TP
.BR \-\-fwupdate\-port
Override device port for the firmware updater

.SH EXAMPLES

.TP
.BR flashgbx " \fR\-\-mode agb \-\-action backup\-rom"
.IP
Creates a backup of the ROM of a Game Boy Advance cartridge

.TP
.BR flashgbx " \fR\-\-mode dmg \-\-action backup\-save"
.IP 
Creates a backup of the save data from a Game Boy cartridge

.TP
.BR flashgbx " \fR\-\-mode agb \-\-action flash\-rom ROM.gba"
.IP
Writes a Game Boy Advance ROM relying on auto\-detecting the flash cartridge

.TP
.BR flashgbx " \fR\-\-mode dmg \-\-action gbcamera\-extract \-\-gbcamera\-outfile\-format png GAMEBOYCAMERA.sav"
.IP
Extracts Game Boy Camera pictures as .png files from a save data file



CHANGES.md
# Release notes
### v3.8 (released 2022-04-21)
- Added support for AGB-E05-01 with MSP55LV100G *(thanks EmperorOfTigers)*
- Added support for DV15 with MSP55LV100G *(thanks zvxr)*
- Added support for SD007_T40_64BALL_TSOP28 with 29LV016T *(thanks Wkr)*
- Confirmed support for AGB-E05-03H with M29W128GH *(thanks Wkr)*
- The integrated firmware updater can now also update the firmware of insideGadgets GBxCart RW v1.1, v1.2, XMAS v1.0 and Mini v1.0 device revisions
- Added experimental support for writing ISX files to Game Boy flash cartridges
- Bundles GBxCart RW v1.4 firmware version R34+L6 (changes the way Game Boy cartridges are read and enables support for some more flash cartridges)
- Added support for the BLAZE Xploder GB unlicensed cheat cartridge (requires firmware version L6+)
- Added support for the unlicensed Sachen mapper (requires firmware version L6+)
- Added support for insideGadgets 128 KB flash cartridges *(thanks AlexiG)*
- Added support for insideGadgets 256 KB flash cartridges *(thanks AlexiG)*
- Minor bug fixes and improvements

### v3.7 (released 2022-03-31)
- Updated the Game Boy Advance lookup database for save types, ROM sizes and checksums (improves support for Classic NES Series, NES Classics and Famicom Mini cartridges)
- When writing new ROMs to Nintendo Power GB Memory Cartridges (DMG-MMSA-JPN), hidden sector data will now be auto-generated if it’s not provided by the user in form of a .map file
- Erasing save data of Game Boy Advance cartridges with the 512K FLASH and 1M FLASH save types is now faster
- Windows Setup Package: Updated the CH340/CH341 driver to the latest version (01/18/2022, 3.7.2022.01)
- Fixed a bug introduced in v3.4 that broke writing ROMs to SD008-6810-V5 with MX29CL256FH cartridges
- Added support for AGB-E05-01 with S29GL064 *(thanks DevDavisNunez)*
- Added support for DIY carts with MBC1 and AM29F010 @ AUDIO *(thanks JS7457)*
- Added support for DIY carts with MBC1 and AM29F040 @ AUDIO *(thanks JS7457)*
- Added support for DIY carts with MBC1 and AM29F080 @ AUDIO *(thanks Timville)*
- Added support for DIY carts with MBC1 and AT49F040 @ AUDIO *(thanks Timville)*
- Added support for DIY carts with MBC1 and SST39SF040 @ AUDIO *(thanks Timville)*
- Added support for B100 with MX29LV640ET *(thanks Mr_V)*
- Added support for B54 with MX29LV320ET *(thanks Mr_V)*
- Added support for AGB-E05-02 with S29GL032 *(thanks redalchemy)*
- Minor bug fixes and improvements

### v3.6 (released 2022-03-09)
- When opening Game Boy Camera Album Viewer manually, the save data will now automatically be loaded if a Game Boy Camera or Pocket Camera cartridge is connected
- Fixed a bug with the insideGadgets 4 MB (S29GL032M) cartridge *(thanks frarees)*
- Fixed a “MemoryError” bug when writing a smaller ROM to the 100BS6600_48BALL_V4 with 6600M0U0BE cartridge with full chip erase mode enabled
- Added support for Flash Advance Card 64M with 28F640J3A120 *(thanks manuelcm1)*
- Fixed a bug with writing ROMs to insideGadgets rumble cartridges on the GBxCart RW v1.3 hardware running the L1 firmware *(thanks DevDavisNunez)*
- In GUI mode, flashable cartridges can now completely be wiped of their ROM data by not selecting any file when using the Write ROM feature *(thanks daidianren for the suggestion)*
- Fixed a bug introduced in v3.4 with making a ROM backup of Nintendo Power GB Memory Cartridges (DMG-MMSA-JPN)
- Bundles GBxCart RW v1.4 firmware version R34+L5 (enables support for some more flash cartridges)
- Added support for DIY cartridges with the AT49F040 flash chip for up to 512 KB of ROM data (requires at least GBxCart RW revision v1.4 and firmware version R34+L5) *(thanks Timville)*
- Added support for the insideGadgets 512 KB flash cartridge (requires at least GBxCart RW revision v1.4 and firmware version R34+L5) *(thanks Timville and marv17)*
- Added support for 100SOP with MSP55LV100S *(thanks Jayro)*

### v3.5 (released 2022-02-14)
- Fixed a bug related to the 100BS6600_48BALL_V4 with 6600M0U0BE cartridge when using the Detect Flash Cart button *(thanks to redalchemy)*
- Fixed a bug with the full chip erase feature via sector erase mode
- When reading save data from a Game Boy Camera or Pocket Camera cartridge via custom command line switches, you will no longer be asked if you’d also like to extract pictures. This is to make it easier to use FlashGBX for scripting. You can still use `--action gbcamera-extract` to extract pictures in a second step, or use the interactive mode.

### v3.4 (released 2022-02-04)
- Confirmed support for SD007_TSOP_64BALL_SOJ28 with unlabeled flash chip *(thanks x7l7j8cc)*
- Added support for SD008-6810-512S with MSP55LV512
- Added support for SD007_TSOP_48BALL_V10 with 29DL164BE-70P *(thanks Herax)*
- Added support for the unlicensed Wisdom Tree mapper *(thanks AlexiG and Smelly-Ghost)*
- Added support for 4000L0ZBQ0 DRV with 3000L0YBQ0 *(thanks iamevn)*
- Added support for SD007_T40_64BALL_TSOP28 with TC58FVB016FT-85 *(thanks joyrider3774)* (cannot be auto-detected, select cartridge type manually)
- Added support for SD007_TSOP_64BALL_SOJ28 with 29DL164BE-70P *(thanks Herax)*
- Confirmed support for Development AGB Cartridge 64M Flash S, E201843
- Added support for the DMG-MBC5-32M-FLASH Development Cartridge, E201264 (requires GBxCart RW firmware L1 or higher)
- If the full chip erase preference is enabled but only sector erase mode is available, the entire chip will now still be erased, so no previous data will remain

### v3.3 (released 2022-01-10)
- Added support for Flash2Advance 128M with 2× 28F640J3A120 *(thanks twitnic)*
- Added support for 36VF3204 and ALTERA CPLD (no PCB text) *(thanks Herax and Davidish)*
- Added support for SD008-6810-V4 with unlabeled flash chip *(thanks LucentW)*
- Will now check written save data for errors if the verification option is enabled (requires GBxCart RW firmware L1 or higher)
- Added a zoom factor setting for extracted Game Boy Camera pictures

FlashGBX/FlashGBX.py
# -*- coding: utf-8 -*-
# FlashGBX
# Author: Lesserkuma (github.com/lesserkuma)

import sys, os, glob, re, json, zlib, argparse, zipfile, traceback, platform, datetime
from . import Util

def ReadConfigFiles(args):
	reset = args['argparsed'].reset
	settings = Util.IniSettings(path=args["config_path"] + "/settings.ini")
	config_version = settings.value("ConfigVersion")
	if not os.path.exists(args["config_path"]): os.makedirs(args["config_path"])
	fc_files = glob.glob("{0:s}/fc_*.txt".format(args["config_path"]))
	if config_version is not None and len(fc_files) == 0:
		print("No flash cartridge type configuration files found in {:s}. Resetting configuration...".format(args["config_path"]))
		settings.clear()
		os.rename(args["config_path"] + "/settings.ini", args["config_path"] + "/settings.ini_" + datetime.datetime.now().strftime("%Y%m%d%H%M%S") + ".bak")
		config_version = False # extracts the config.zip again
	elif reset:
		settings.clear()
		print("All configuration has been reset.")
	
	settings.setValue("ConfigVersion", Util.VERSION)
	return (config_version, fc_files)

def LoadConfig(args):
	app_path = args['app_path']
	config_path = args['config_path']
	ret = []
	flashcarts = { "DMG":{}, "AGB":{} }
	
	# Settings and Config
	(config_version, fc_files) = ReadConfigFiles(args=args)
	if config_version != Util.VERSION:
		# Rename old files that have since been replaced/renamed/merged
		deprecated_files = [ "fc_AGB_TEST.txt", "fc_DMG_TEST.txt", "fc_AGB_Nintendo_E201850.txt", "fc_AGB_Nintendo_E201868.txt", "config.ini", "fc_DMG_MX29LV320ABTC.txt", "fc_DMG_iG_4MB_MBC3_RTC.txt", "fc_AGB_Flash2Advance.txt", "fc_AGB_MX29LV640_AUDIO.txt", "fc_AGB_M36L0R7050T.txt", "fc_AGB_M36L0R8060B.txt", "fc_AGB_M36L0R8060T.txt", "fc_AGB_iG_32MB_S29GL512N.txt" ]
		for file in deprecated_files:
			if os.path.exists(config_path + "/" + file):
				os.rename(config_path + "/" + file, config_path + "/" + file + "_" + datetime.datetime.now().strftime("%Y%m%d%H%M%S") + ".bak")
		
		rf_list = ""
		if os.path.exists(app_path + "/res/config.zip"):
			with zipfile.ZipFile(app_path + "/res/config.zip") as zip:
				for zfile in zip.namelist():
					if os.path.exists(config_path + "/" + zfile):
						zfile_crc = zip.getinfo(zfile).CRC
						with open(config_path + "/" + zfile, "rb") as ofile: buffer = ofile.read()
						ofile_crc = zlib.crc32(buffer) & 0xFFFFFFFF
						if zfile_crc == ofile_crc: continue
						os.rename(config_path + "/" + zfile, config_path + "/" + zfile + "_" + datetime.datetime.now().strftime("%Y%m%d%H%M%S") + ".bak")
						rf_list += zfile + "\n"
					zip.extract(zfile, config_path + "/")
			
			if rf_list != "":
				ret.append([1, "The application was recently updated and some flashcart type files have been updated as well. You will find backup copies of them in your configuration directory.\n\nUpdated files:\n" + rf_list[:-1]])
			fc_files = glob.glob("{0:s}/fc_*.txt".format(config_path))
		else:
			print("WARNING: {:s} not found. This is required to load new flash cartridge type configurations after updating.".format(app_path + "/res/config.zip"))
	
	# Read flash cart types
	for file in fc_files:
		with open(file, encoding='utf-8') as f:
			data = f.read()
			specs_int = re.sub("(0x[0-9A-F]+)", lambda m: str(int(m.group(1), 16)), data) # hex numbers to int numbers, otherwise not valid json
			try:
				specs = json.loads(specs_int)
			except:
				ret.append([2, "The flashchip type file “{:s}” could not be parsed and needs to be fixed before it can be used.".format(os.path.basename(file))])
				continue
			for name in specs["names"]:
				if not specs["type"] in flashcarts: continue # only DMG and AGB are supported right now
				flashcarts[specs["type"]][name] = specs
	
	return { "flashcarts":flashcarts, "config_ret":ret }

class ArgParseCustomFormatter(argparse.ArgumentDefaultsHelpFormatter, argparse.RawDescriptionHelpFormatter): pass
def main(portableMode=False):
	if platform.system() == "Windows": os.system("color")
	os.environ['QT_MAC_WANTS_LAYER'] = '1'
	
	print("{:s} {:s} by Lesserkuma".format(Util.APPNAME, Util.VERSION))
	print("https://github.com/lesserkuma/FlashGBX")
	#print("\nDISCLAIMER: This software is provided as-is and the developer is not responsible for any damage that is caused by the use of it. Use at your own risk!")
	
	if getattr(sys, 'frozen', False) and hasattr(sys, '_MEIPASS'):
		app_path = os.path.dirname(sys.executable)
	else:
		app_path = os.path.dirname(os.path.abspath(__file__))
	
	try:
		from PySide2 import QtCore
		cp = { "subdir":app_path + "/config", "appdata":QtCore.QStandardPaths.writableLocation(QtCore.QStandardPaths.AppConfigLocation) + '/FlashGBX' }
	except:
		cp = { "subdir":app_path + "/config", "appdata":os.path.expanduser('~') + '/FlashGBX' }
	
	if portableMode:
		cfgdir_default = "subdir"
	else:
		cfgdir_default = "appdata"
	
	examples = "\nexamples:\n" + \
	"  Backup the ROM of a Game Boy Advance cartridge:\n\tFlashGBX --mode agb --action backup-rom\n\n" + \
	"  Backup Save Data from a Game Boy cartridge:\n\tFlashGBX --mode dmg --action backup-save\n\n" + \
	"  Write a Game Boy Advance ROM relying on auto-detecting the flash cartridge:\n\tFlashGBX --mode agb --action flash-rom ROM.gba\n\n" + \
	"  Extract Game Boy Camera pictures as .png files from a save data file:\n\tFlashGBX --mode dmg --action gbcamera-extract --gbcamera-outfile-format png GAMEBOYCAMERA.sav\n\n"

	parser = argparse.ArgumentParser(formatter_class=ArgParseCustomFormatter, epilog=examples)
	try:
		parser._action_groups[1].title = "general arguments"
	except:
		pass
	parser.add_argument("--cli", help="force command line interface mode", action="store_true")
	parser.add_argument("--reset", help="clears all settings such as last used directory information", action="store_true")
	parser.add_argument("--debug", help="enable debug messages used for development", action="store_true")
	
	parser.add_argument_group('')
	ap_config = parser.add_argument_group('configuration arguments')
	if "appdata" in cp: ap_config.add_argument("--cfgdir", choices=["appdata", "subdir"], type=str.lower, default=cfgdir_default, help="sets the config directory to either the OS-provided local app config directory (" + cp['appdata'] + "), or a subdirectory of this application (" + cp['subdir'].replace("\\", "/") + ")")
	
	ap_cli1 = parser.add_argument_group('main command line interface arguments')
	ap_cli1.add_argument("--mode", choices=["dmg", "agb"], type=str.lower, default=None, help="set cartridge mode to \"dmg\" (Game Boy) or \"agb\" (Game Boy Advance)")
	ap_cli1.add_argument("--action", choices=["info", "backup-rom", "flash-rom", "backup-save", "restore-save", "erase-save", "gbcamera-extract", "fwupdate-gbxcartrw", "debug-test-save"], type=str.lower, default=None, help="select program action")
	ap_cli1.add_argument("--overwrite", action="store_true", help="overwrite without asking if target file already exists")
	ap_cli1.add_argument("path", nargs="?", default="auto", help="target or source file path (optional when reading, required when writing)")
	
	ap_cli2 = parser.add_argument_group('optional command line interface arguments')
	ap_cli2.add_argument("--dmg-romsize", choices=["auto", "32kb", "64kb", "128kb", "256kb", "512kb", "1mb", "2mb", "4mb", "8mb", "16mb", "32mb"], type=str.lower, default="auto", help="set size of Game Boy cartridge ROM data")
	ap_cli2.add_argument("--dmg-mbc", choices=["auto", "1", "2", "3", "5", "6", "7"], type=str.lower, default="auto", help="set memory bank controller type of Game Boy cartridge")
	ap_cli2.add_argument("--dmg-savesize", choices=["auto", "4k", "16k", "64k", "256k", "512k", "1m", "eeprom2k", "eeprom4k", "tama5", "4m"], type=str.lower, default="auto", help="set size of Game Boy cartridge save data")
	ap_cli2.add_argument("--agb-romsize", choices=["auto", "1mb", "2mb", "4mb", "8mb", "16mb", "32mb", "64mb", "128mb", "256mb"], type=str.lower, default="auto", help="set size of Game Boy Advance cartridge ROM data")
	ap_cli2.add_argument("--agb-savetype", choices=["auto", "eeprom4k", "eeprom64k", "sram256k", "flash512k", "flash1m", "dacs8m", "sram512k", "sram1m"], type=str.lower, default="auto", help="set type of Game Boy Advance cartridge save data")
	ap_cli2.add_argument("--store-rtc", action="store_true", help="store RTC register values if supported")
	ap_cli2.add_argument("--ignore-bad-header", action="store_true", help="don’t stop if invalid data found in cartridge header data")
	ap_cli2.add_argument("--flashcart-type", type=str, default="autodetect", help="name of flash cart; see txt files in config directory")
	ap_cli2.add_argument("--prefer-chip-erase", action="store_true", help="prefer full chip erase over sector erase when both available")
	ap_cli2.add_argument("--reversed-sectors", action="store_true", help="use reversed flash sectors if possible")
	ap_cli2.add_argument("--force-5v", action="store_true", help="force 5V when writing Game Boy flash cartridges")
	ap_cli2.add_argument("--no-verify-write", action="store_true", help="do not verify written data")
	ap_cli2.add_argument("--save-filename-add-datetime", action="store_true", help="adds a timestamp to the file name of save data backups")
	ap_cli2.add_argument("--gbcamera-palette", choices=["grayscale", "dmg", "sgb", "cgb1", "cgb2", "cgb3"], type=str.lower, default="grayscale", help="sets the palette of pictures extracted from Game Boy Camera saves")
	ap_cli2.add_argument("--gbcamera-outfile-format", choices=["png", "bmp", "gif", "jpg"], type=str.lower, default="png", help="sets the file format of saved pictures extracted from Game Boy Camera saves")
	ap_cli2.add_argument("--fwupdate-port", help="override device port for the firmware updater", default=None)
	args = parser.parse_args()
	
	if "appdata" in cp:
		config_path = cp[args.cfgdir]
	else:
		config_path = cp["subdir"]
	
	if args.mode is not None or args.action is not None:
		args.cli = True
	
	if args.debug == True:
		Util.DEBUG = True
	
	args = {"app_path":app_path, "config_path":config_path, "argparsed":args}
	args.update(LoadConfig(args))
	
	app = None
	exc = None
	if not args["argparsed"].cli:
		try:
			from . import FlashGBX_GUI
			app = FlashGBX_GUI.FlashGBX_GUI(args)
		except ModuleNotFoundError:
			app = None
		except:
			exc = traceback.format_exc()
			app = None
		
		if app is None:
			from . import FlashGBX_CLI
			if args["argparsed"].action is None:
				if exc is not None: print("{:s}{:s}{:s}".format(Util.ANSI.YELLOW, exc, Util.ANSI.RESET))
				parser.print_help()
				print("\n\n{:s}NOTE: GUI mode couldn’t be launched, but the application can be run in CLI mode.\n      Optional command line switches are explained above.{:s}\n".format(Util.ANSI.RED, Util.ANSI.RESET))
			
			print("Now running in CLI mode.\n")
			app = FlashGBX_CLI.FlashGBX_CLI(args)
			try:
				app.run()
			except KeyboardInterrupt:
				print("\n\nProgram stopped.")
			return
		
		app.run()
	
	else:
		from . import FlashGBX_CLI
		print("Now running in CLI mode.\n")
		app = FlashGBX_CLI.FlashGBX_CLI(args)
		try:
			app.run()
		except KeyboardInterrupt:
			print("\n\nProgram stopped.")
